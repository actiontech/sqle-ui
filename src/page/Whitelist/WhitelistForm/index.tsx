import { Form, Input } from 'antd';
import { useTranslation } from 'react-i18next';
import MonacoEditor from 'react-monaco-editor';
import { ModalFormLayout } from '../../../data/common';
import useChangeTheme from '../../../hooks/useChangeTheme';
import useStyles from '../../../theme';
import { WhitelistFormProps } from './index.type';

const WhitelistForm: React.FC<WhitelistFormProps> = (props) => {
  const { t } = useTranslation();
  const styles = useStyles();
  const { currentEditorTheme } = useChangeTheme();

  return (
    <Form form={props.form} {...ModalFormLayout}>
      <Form.Item label={t('whitelist.table.desc')} name="desc">
        <Input.TextArea
          className="textarea-no-resize"
          autoSize={{
            minRows: 3,
            maxRows: 10,
          }}
          placeholder={t('common.form.placeholder.input')}
        />
      </Form.Item>
      <Form.Item
        name="sql"
        label={t('order.sqlInfo.sql')}
        initialValue="/* input your sql */"
        wrapperCol={{
          ...ModalFormLayout.wrapperCol,
          className: styles.editor,
        }}
        rules={[
          {
            required: true,
          },
        ]}
      >
        <MonacoEditor
          theme={currentEditorTheme}
          width="100%"
          height="500"
          language="sql"
        />
      </Form.Item>
    </Form>
  );
};

export default WhitelistForm;
