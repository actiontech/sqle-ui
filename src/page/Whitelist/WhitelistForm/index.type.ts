import { FormInstance } from 'antd';

export type WhitelistFormFields = {
  desc?: string;
  sql: string;
};

export type WhitelistFormProps = {
  form: FormInstance<WhitelistFormFields>;
};
