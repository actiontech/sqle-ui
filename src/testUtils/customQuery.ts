import * as domTestingLib from '@testing-library/dom';
import { fireEvent, screen } from '@testing-library/react';

export const getBySelector = (selector: string, baseElement?: Element) => {
  let temp;
  if (baseElement !== undefined) {
    temp = baseElement.querySelectorAll(selector);
  } else {
    temp = document.querySelectorAll(selector);
  }
  if (temp.length === 0) {
    throw domTestingLib.queryHelpers.getElementError(
      `Unable to find an element by: selector=${selector}`,
      document.body
    );
  }
  if (temp.length > 1) {
    throw domTestingLib.queryHelpers.getElementError(
      `Found multiple elements with the selector=${selector}`,
      document.body
    );
  }
  return temp[0];
};

export const getAllBySelector = (selector: string, baseElement?: Element) => {
  let temp;
  if (baseElement !== undefined) {
    temp = baseElement.querySelectorAll(selector);
  } else {
    temp = document.querySelectorAll(selector);
  }
  if (temp.length === 0) {
    throw domTestingLib.queryHelpers.getElementError(
      `Unable to find an element by: selector=${selector}`,
      document.body
    );
  }
  return temp;
};

export const getSelectContentByFormLabel = (label: string) => {
  return getBySelector(
    '.ant-select-selection-item-content',
    screen.getByText(label).parentNode?.parentNode as HTMLDivElement
  );
};

export const getSelectValueByFormLabel = (label: string) => {
  return getBySelector(
    '.ant-select-selection-item',
    screen.getByLabelText(label).parentNode?.parentNode as HTMLDivElement
  );
};

/**
 * 模拟选择 Select 组件中下拉框中的某项数据
 * @param 表单中 Select 组件外部 Form.Item 的label, 注意该 Form.Item 必须有 name
 * @param optionText 需要选中的下拉项
 * @param index 通过文本获取所有的节点会有两个(具体信息可以在网页上查看Select 组件中下拉的节点信息), 正常来说其中第二个是需要来需要用户点击的,
 * 所以这里的 index 默认值为 1. 并且这个节点拥有类名 ant-select-item-option-content, 所以在点击前需要使用 expect 来判断是否为正确的节点.
 */
export const selectOptionByIndex = (
  label: string,
  optionText: string,
  index = 1
) => {
  fireEvent.mouseDown(screen.getByLabelText(label));
  const options = screen.getAllByText(optionText);
  let realIndex = index;
  if (index < 0) {
    realIndex = options.length + index;
  }
  const option = options[realIndex];
  expect(option).toHaveClass('ant-select-item-option-content');
  fireEvent.click(option);
};

/**
 * 具体功能同 selectOptionByIndex, 不过该函数主要针对下拉框选项为一个节点的情况
 * @param label 表单中 Select 组件外部 Form.Item 的label, 注意该 Form.Item 必须有 name
 * @param optionCls 下拉项节点最外层的 class
 * @param index  这里不同于 selectOptionByIndex 的 index, 其代表的就是真实需要选择的下拉项的序号, 从 0 开始.
 */
export const selectCustomOptionByClassName = (
  label: string,
  optionCls: string,
  index: number
) => {
  fireEvent.mouseDown(screen.getByLabelText(label));
  const options = getAllBySelector(`.${optionCls}`);
  let realIndex = index;
  if (index < 0) {
    realIndex = options.length + index;
  }

  const option = options[realIndex];
  expect(option.parentNode).toHaveClass('ant-select-item-option-content');
  fireEvent.click(option);
};

export const getHrefByText = (name: string) => {
  return screen.getByText(name).getAttribute('href');
};

export const getAllHrefByText = (name: string) => {
  return screen.getAllByText(name).map((v) => v.getAttribute('href'));
};
