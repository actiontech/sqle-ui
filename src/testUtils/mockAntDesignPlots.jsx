import { cloneDeep, isObject } from 'lodash';
const MockPlots = (props) => {
  const cloneProps = cloneDeep(props);

  Object.keys(cloneProps).forEach((key) => {
    if (isObject(cloneProps[key])) {
      cloneProps[key] = JSON.stringify(cloneProps[key]);
    }
  });
  return <div {...cloneProps} />;
};

const Line = MockPlots;
const Pie = MockPlots;
const RadialBar = MockPlots;
const Gauge = MockPlots;
const Column = MockPlots;
const mockRegisterShape = jest.fn();
const G2 = {
  registerShape: jest.fn(),
};

export { Line, Pie, RadialBar, Gauge, Column, G2, mockRegisterShape };

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  Line,
  Pie,
  RadialBar,
  Gauge,
  Column,
  G2,
};
