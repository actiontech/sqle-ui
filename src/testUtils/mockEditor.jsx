const mockEditor = (props) => {
  // https://github.com/react-monaco-editor/react-monaco-editor/issues/176
  const { editorDidMount, ...otherProps } = props;
  return <input {...otherProps} />;
};

export default mockEditor;
