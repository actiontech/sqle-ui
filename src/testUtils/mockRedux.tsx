import { Dictionary } from '../types/common.type';
import * as redux from 'react-redux';
import locale from '../store/locale';
import user from '../store/user';
import userManage from '../store/userManage';
import whitelist from '../store/whitelist';
import nav from '../store/nav';
import { combineReducers, createStore } from 'redux';
import React from 'react';
import { Provider } from 'react-redux';

const reducers = combineReducers({
  locale,
  user,
  userManage,
  whitelist,
  nav,
});

export const storeFactory = (initStore: Dictionary = {}) => {
  return createStore(reducers, initStore);
};

export const mockUseSelector = (store?: any) => {
  const spy = jest.spyOn(redux, 'useSelector');
  spy.mockImplementation((getFn: Function) => {
    return getFn(store);
  });
  return spy;
};

export const mockUseDispatch = () => {
  const scopeDispatch = jest.fn();
  const spy = jest.spyOn(redux, 'useDispatch');
  spy.mockImplementation(() => scopeDispatch);
  return { spy, scopeDispatch };
};

export const mockUseLocation = () => {}


export const CustomProvider: React.FC<{ initStore: Dictionary }> = (props) => {
  return (
    <Provider store={storeFactory(props.initStore)}>{props.children}</Provider>
  );
};
