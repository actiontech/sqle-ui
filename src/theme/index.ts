import { makeStyles } from '@material-ui/styles';
import { Theme } from '../types/theme.type';
import darkTheme from './dark';
import lightTheme from './light';

enum SupportTheme {
  DARK = 'dark',
  LIGHT = 'light',
}

const ThemeData = {
  [SupportTheme.DARK]: darkTheme,
  [SupportTheme.LIGHT]: lightTheme,
};

const useStyles = makeStyles<Theme>((theme) => ({
  headerBg: {
    backgroundColor: theme.header.background,
    color: theme.header.color,
  },
  editor: {
    border: theme.editor.border,
    paddingRight: 1,
  },
}));

export { SupportTheme, ThemeData };
export default useStyles;
