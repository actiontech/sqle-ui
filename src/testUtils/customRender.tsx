import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import {
  BrowserRouter,
  Router,
  RouterProps,
  MemoryRouter,
  MemoryRouterProps,
} from 'react-router-dom';
import { Provider } from 'react-redux';
import { Dictionary } from '../types/common.type';
import { storeFactory } from './mockRedux';
import lightTheme from '../theme/light';
import { ThemeProvider } from '@material-ui/styles';
import { mount, shallow } from 'enzyme';

type RenderParams = Parameters<typeof render>;
type MountParams = Parameters<typeof mount>;
type ShallowParams = Parameters<typeof shallow>;

export const renderWithRouter = (...[ui, option]: [...RenderParams]) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>, option);
};

export const renderWithRedux = (
  ...[ui, option, initStore]: [...RenderParams, Dictionary?]
) => {
  return render(
    <Provider store={storeFactory(initStore)}>{ui}</Provider>,
    option
  );
};

export const renderHooksWithRedux = <TProps, TResult>(
  hooks: (props: TProps) => TResult,
  storeState: Dictionary = {}
) => {
  return renderHook(hooks, {
    wrapper: Provider,
    initialProps: {
      store: storeFactory(storeState),
    },
  } as any);
};

export const renderWithServerRouter = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(<Router {...props}>{ui}</Router>, option);
};

export const renderWithMemoryRouter = (
  ...[ui, option, props]: [...RenderParams, MemoryRouterProps?]
) => {
  return render(<MemoryRouter {...props}>{ui}</MemoryRouter>, option);
};

export const renderWithTheme = (...[ui, option]: [...RenderParams]) => {
  return render(<ThemeProvider theme={lightTheme}>{ui}</ThemeProvider>, option);
};

export const renderWithThemeAndRouter = (
  ...[ui, option]: [...RenderParams]
) => {
  return render(
    <MemoryRouter>
      <ThemeProvider theme={lightTheme}>{ui}</ThemeProvider>
    </MemoryRouter>,
    option
  );
};

export const mountWithTheme = (...[ui, option]: [...MountParams]) => {
  return mount(ui, {
    ...option,
    wrappingComponent: ThemeProvider,
    wrappingComponentProps: {
      theme: lightTheme,
    },
  });
};

export const shallowWithTheme = (...[ui, option]: [...ShallowParams]) => {
  return shallow(ui, {
    ...option,
    wrappingComponent: ThemeProvider,
    wrappingComponentProps: {
      theme: lightTheme,
    },
  });
};
