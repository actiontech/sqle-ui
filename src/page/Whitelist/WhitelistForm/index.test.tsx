import { renderHook } from '@testing-library/react-hooks';
import { useForm } from 'antd/lib/form/Form';
import WhitelistForm from '.';
import { renderWithTheme } from '../../../testUtils/customRender';
import { mockUseDispatch, mockUseSelector } from '../../../testUtils/mockRedux';
import { SupportTheme } from '../../../theme';
// https://github.com/react-monaco-editor/react-monaco-editor/issues/176
jest.mock('react-monaco-editor', () => {
  return (props: any) => <input {...props} />;
});

describe('Whitelist/WhitelistForm', () => {
  beforeEach(() => {
    mockUseDispatch();
    mockUseSelector({ user: { theme: SupportTheme.LIGHT } });
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  test('should match snapshot', () => {
    const { result } = renderHook(() => useForm());
    const { container } = renderWithTheme(
      <WhitelistForm form={result.current[0]} />
    );
    expect(container).toMatchSnapshot();
    expect(Object.keys(result.current[0].getFieldsValue())).toEqual([
      'desc',
      'sql',
    ]);
    expect(result.current[0].getFieldsValue()).toEqual({
      sql: '/* input your sql */',
    });
  });
});
