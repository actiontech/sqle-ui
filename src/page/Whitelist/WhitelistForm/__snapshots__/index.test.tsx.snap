// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Whitelist/WhitelistForm should match snapshot 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-7"
      >
        <label
          class=""
          for="matchType"
          title="whitelist.table.matchType"
        >
          whitelist.table.matchType
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control ant-col-xs-24 ant-col-sm-13"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-radio-group ant-radio-group-outline"
              id="matchType"
            >
              <label
                class="ant-radio-wrapper ant-radio-wrapper-checked"
              >
                <span
                  class="ant-radio ant-radio-checked"
                >
                  <input
                    checked=""
                    class="ant-radio-input"
                    type="radio"
                    value="exact_match"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  whitelist.matchType.exact
                </span>
              </label>
              <label
                class="ant-radio-wrapper"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    type="radio"
                    value="fp_match"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  whitelist.matchType.fingerPrint
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-7"
      >
        <label
          class=""
          for="desc"
          title="whitelist.table.desc"
        >
          whitelist.table.desc
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control ant-col-xs-24 ant-col-sm-13"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <textarea
              class="ant-input textarea-no-resize"
              id="desc"
              placeholder="common.form.placeholder.input"
              style="height: -40px; min-height: -12px; max-height: -40px; overflow-y: hidden; resize: none; overflow-x: hidden;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-7"
      >
        <label
          class="ant-form-item-required"
          for="sql"
          title="order.sqlInfo.sql"
        >
          order.sqlInfo.sql
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control makeStyles-editor-3 ant-col-xs-24 ant-col-sm-13"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <input
              height="220"
              id="sql"
              language="sql"
              theme="vs"
              value="/* input your sql */"
              width="100%"
            />
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
